nixos/qemu-vm: Allow to build and run VMs on Darwin

YorikSar opened the pull request:
###### Description of changes

Introduce an option `virtualisation.qemu.pkgs` that allows to set a package set that should be used to run VM (port of a commit by @infinisil). It allows you to specify, for example, `aarch64-darwin` package set to run `aarch64-linux` configuration in VM with native virtualisation on macOS.

Also add Qemu parameters for running VMs with HVF acceleration on both x86_64 and aarch64.

Related to https://github.com/NixOS/nixpkgs/issues/108984.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

infinisil committed:
nixos/qemu: Add pkgs option for allowing other systems to run the VM

YorikSar committed:
nixos/qemu: nixpkgs-fmt nixos/lib/qemu-common.nix

YorikSar committed:
nixos/qemu: Add QEMU arguments for HVF on Darwin

Only support the same architechture for now, no emulation.

Related to #108984.

roberth commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -558,10 +560,22 @@ in
       };
 
     virtualisation.qemu = {
+      pkgs = mkOption {
```
This is not a property of qemu, but of the host. I'd suggest `virtualisation.host.pkgs` instead of `virtualisation.qemu.pkgs`. It matches the description better, and also aligns with `hostPkgs` in https://github.com/NixOS/nixpkgs/pull/176557

-----

_Thoughts_

The homonymy with `nixpkgs.hostPlatform` is unfortunate, but I don't think we can use another term than `host`. `hypervisor` is too specific (it would be nice for the option to be usable in a container-based setting). `monitor` is just worse.
At least the `virtualisation` prefix gives a clue to resolve the ambiguity, so it seems that `virtualisation.host.pkgs` is the best we can do.

(As a side note, arguably the host isn't part of the system, but it's a little late to backtrack on the `virtualisation.qemu.package` option, which is also a host property.)

YorikSar replied:
I'm not sure we should break consistency with `virtualisation.qemu.package` here. `virtualisation` namespace as it is is quite a mess of both "this config as a host" and "this config as a guest" options. I don't think adding another one directly there would be better than keeping everything related to qemu-vm module to `virtualisation.qemu`. Besides, this package set is used only for building and wrapping the qemu binary itself.

If we want to accommodate possibility of building other artefacts from the config (like Docker image, maybe?), I think we should have this option somewhere in a more generic place like `system.build.host.pkgs` (although I don't think we can add anything to `system.build`). I think this would be a part of a separate effort to unify such abstraction across all builders, which is not the aim of this pull request.

roberth replied:
Why wait? You'd just create more legacy options.

> I think we should have this option somewhere in a more generic place like `system.build.host.pkgs`

`system.build` is for the actual "outputs" of a config, such as `toplevel` and images.

`host.pkgs` is more of an input, like most options, including `nixpkgs.pkgs`. It just happens to be used in derivations that may produce outputs in `system.build`.

> (although I don't think we can add anything to `system.build`).

I don't think there's a technical limitation. You can even define proper options there nowadays, instead of just config attrs.

YorikSar replied:
I actually didn't know about `nixpkgs.pkgs` at all. It looks like we already have `hostPlatform` and `buildPlatform` there. Maybe we should also have `nixpkgs.targetPlatform` and `nixpkgs.targetPkgs` or something like this? It would keep the terminology consistent with cross-compilation.

Then if I'm on aarch64-darwin laptop want to cross-compile aarch64-linux system on x86_64-linux and run it in VM, I will have:

```
nixpkgs.hostPlatform.system = "aarch64-linux";
nixpkgs.buildPlatform.system = "x86_64-linux";
nixpkgs.targetPlatform.system = "aarch64-darwin";
```

And then we can set `nixpkgs.targetPlatform.system` automatically by `nixos-rebuild build-vm`, for example, and use it for any future additions like `nixos-rebuild build-docker-image`.

I feel like `nixpkgs` would be a better place for this option (or options) than `virtualisation`, but I'm not sure about the naming. The meaning of these terms in cross-compiling space always confuses me.

infinisil replied:
I agree with @roberth on this, and I think his original proposal of `virtualisation.host.pkgs` is a good idea, since it's specific to only virtualisation. We could also rename `virtualisation.qemu.package` to `virtualisation.qemu.host.package` if we want all host options a `host` in their option path.

roberth replied:
Apologies if the first two parts of my reply are redundant (because you seem to agree) but some things should be clarified.

> I didn't mean that `targetPlatform` would be used for stdenv configuration. As you correctly stated, this would make the build result depend on which platform evaluation happens.

I don't think that's what I've stated. Perhaps I should have stressed that the "evaluation platform" is forgotten after instantiation and does not even exist at all in pure mode.

> I didn't mean that `targetPlatform` would be used for stdenv configuration.

Then let's avoid this name. Application of the original `targetPlatform` concept in `stdenv` is confusing enough as it is, so let's not add any ambiguity. A VM is usually not a compiler, and if it is, it is a JIT, so `build`, `host` and `target` are the same. The only thing that isn't the same is the source architecture.

Were you referring to `nixos-rebuild --target-host`? That option has basically nothing to do with `targetPlatform` and should be renamed. It just so happens that _during Nixpkgs bootstrapping_ these terms do line up somewhat, but bootstrapping is abstracted away by Nixpkgs and is not something you should ever have to worry about, unless you're packaging compilers that are part of stdenv.

> That's why I'm suggesting that `nixos-rebuild build-vm` could provide local system to the NixOS configuration in `virtualisation.host.pkgs` or `virtualisation.host.platform.system`, so that the resulting script can run locally on the system that `nixes-rebuild` has been called on.

This is definitely something we should pay attention to. Essentially you're asking for `virtualisation.host` to become impure. This is straightforward in the non-flakes case, and otherwise feasible by making `nixos-rebuild` invoke `extendModules` on the configuration, although I'm not sure if we should.
I don't think we have to implement this functionality for this PR to be useful, so I'd recommend to follow it up later.

infinisil commented on nixos/lib/qemu-common.nix:
```diff
@@ -4,29 +4,41 @@
+      aarch64-linux = "${qemuPkg}/bin/qemu-system-aarch64 -machine virt,gic-version=max,accel=kvm:tcg -cpu max";
+      powerpc64le-linux = "${qemuPkg}/bin/qemu-system-ppc64 -machine powernv";
+      powerpc64-linux = "${qemuPkg}/bin/qemu-system-ppc64 -machine powernv";
+      x86_64-darwin = "${qemuPkg}/bin/qemu-kvm -cpu max";
```
I don't fully get this. Is this the command for emulating a `x86_64-darwin` machine for any Linux qemu? Maybe that could be added as an explanation above, the `isLinux` followed by this `darwin` here is confusing

infinisil commented on nixos/lib/qemu-common.nix:
```diff
@@ -4,29 +4,41 @@
+        if qemuPkg.stdenv.isAarch64 then {
+          aarch64-linux = "${qemuPkg}/bin/qemu-system-aarch64 -machine virt,gic-version=2,accel=hvf:tcg -cpu max";
+        } else if qemuPkg.stdenv.isx86_64 then {
+          x86_64-linux = "${qemuPkg}/bin/qemu-system-x86_64 -machine type=q35,accel=hvf:tcg -cpu max";
```
And I guess this is the other way around then, emulating `x86_64-linux` from any x86_64 darwin.

The fact that this has to be declared this way leads me to believe this is expressed in a rather poor way. What about emulating other systems from darwin? Where do the command line arguments come from? Could we set up a matrix instead of nested if conditions?

YorikSar reviewed (commented):
I've moved `pkgs` parameter to `virtualisation.host.pkgs` and updated host-guest mapping for QEMU parameters.

YorikSar commented:
Something's wrong with my test VM config, will investigate.

YorikSar commented:
Ah, just forgot to apply fix for 9p, everything's good. Sorry for the confusion.

prusnak commented:
What's the easiest way how to test this?

infinisil reviewed (approved):
Looking good! I can merge after conflict is resolved.

YorikSar commented:
@prusnak 
> What's the easiest/minimalistic way how to test this?

You can find my attempts at using this in https://github.com/NixOS/nixpkgs/issues/108984#issuecomment-1150149700 - it uses my temporary branch from which I extracted this PR. Note that it's a bit older than current version of this PR, and it requires you to use your own Linux builder or use the exact flake and my Cachix substituter.

@infinisil 
> Looking good! I can merge after conflict is resolved.
Only nixpkgs-fmt patch conflicted with master. I've rebased on top of the latest master now.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-35/21701/1
